aviatrix/resource_aviatrix_controller_private_mode_config.go

```
@@ -0,0 +1,191 @@
+			},
+			"copilot_instance_id": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Copilot instance ID to associate with the Controller for Private Mode.",
+			},
+			"proxies": {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: "Set of proxies.",
+			},
+		},
+	}
+}
+
+func resourceAviatrixControllerPrivateModeConfigCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*goaviatrix.Client)
+
+	enablePrivateMode := d.Get("enable_private_mode").(bool)
+	if !enablePrivateMode {
+		for _, key := range []string{"copilot_instance_id", "proxies"} {
+			if _, ok := d.GetOk(key); ok {
+				return diag.Errorf("failed to create Controller Private Mode config: %q must be empty when %q is false", key, "enable_private_mode")
+			}
+		}
+	}
+
+	flag := false
+	defer resourceAviatrixControllerPrivateModeConfigReadIfRequired(ctx, d, meta, &flag)
+
+	if enablePrivateMode {
+		err := client.EnablePrivateMode(ctx)
+		if err != nil {
+			return diag.Errorf("failed to enable Private Mode: %s", err)
+		}
+
+		if _, ok := d.GetOk("copilot_instance_id"); ok {
+			copilotInstanceId := d.Get("copilot_instance_id").(string)
+			err = client.UpdatePrivateModeCopilot(ctx, copilotInstanceId)
+			if err != nil {
+				return diag.Errorf("failed to set Copilot instance ID: %s", err)
+			}
+		}
+
+		if _, ok := d.GetOk("proxies"); ok {
+			proxiesSet := d.Get("proxies").(*schema.Set)
+			proxies := make([]string, proxiesSet.Len())
+			for i, proxy := range proxiesSet.List() {
+				proxies[i] = proxy.(string)
+			}
+
+			err = client.UpdatePrivateModeControllerProxies(ctx, proxies)
+			if err != nil {
+				return diag.Errorf("failed to set Controller proxies: %s", err)
+			}
+		}
+	} else {
+		err := client.DisablePrivateMode(ctx)
+		if err != nil {
+			return diag.Errorf("failed to disable Private Mode: %s", err)
+		}
+	}
+
```

Do we need to return the error if private mode is already disabled? Should we just log it like controller_security_group_management?
@zhiqiangjin

---

We should be consistent with this type of error.

---

If Private Mode is already disabled, the `DisablePrivateMode` method will not return an error. I don't think this needs a change.